File: backend/app/services/playbook.py
from __future__ import annotations

from typing import Any


def playbook_config(twin: dict[str, Any]) -> dict[str, Any]:
    """Map a MarketTwin result to a copy-pasteable Bitget GetAgent Playbook block.

    Keeps AlphaArena paper-only: output is a monitoring/pause config suggestion,
    never an order. Field names mirror Playbook: Strategy, Pair, Amount, Frequency,
    Entry Condition, Pause Condition, Market Regime, Signal Confidence, Optimizer Reason.
    """
    shock = twin.get("shock") or {}
    impacts = twin.get("impacts") or []
    top = sorted(impacts, key=lambda r: abs(float(r.get("impact_pct") or 0)), reverse=True)[:3]
    pairs = [str(r.get("symbol")) for r in top]
    worst = min([float(r.get("impact_pct") or 0) for r in top], default=0.0)
    return {
        "Strategy": f"Stress-watch: {shock.get('driver', 'custom shock')} {shock.get('direction', '')} {shock.get('magnitude', '')}{shock.get('unit', '')}",
        "Pair": ", ".join(pairs) if pairs else "rNVDA",
        "Amount per Cycle": "Paper only in AlphaArena — set your Playbook paper size",
        "Frequency": str(twin.get("duration") or "24H"),
        "Entry Condition": f"Enter review when driver moves {shock.get('magnitude', '')}{shock.get('unit', '')} {shock.get('direction', '')}; most sensitive: {pairs[0] if pairs else 'rNVDA'}",
        "Pause Condition": f"Pause if realized stress exceeds worst estimated {worst:.2f}% or invalidation conditions trigger",
        "Market Regime": str(shock.get("category", "nasdaq")),
        "Signal Confidence": f"{twin.get('model_source', 'deterministic')} | calibrated={any(bool(r.get('calibrated')) for r in top)}",
        "Optimizer Reason": str((twin.get("explanation_view") or {}).get("impact_summary") or twin.get("explanation") or "")[:500],
        "Source": "AlphaArena MarketTwin (paper-only suggestion, not an order)",
    }

File: src/lib/playbook.ts
import type { TwinResponse } from '../product/api';

export function twinToPlaybookText(result: TwinResponse): string {
  const top = [...result.impacts]
    .sort((a, b) => Math.abs(b.impact_pct) - Math.abs(a.impact_pct))
    .slice(0, 3);
  const pairs = top.map((i) => i.symbol).join(', ') || 'rNVDA';
  const worst = top.length ? Math.min(...top.map((i) => i.impact_pct)) : 0;
  const lines: Array<[string, string]> = [
    ['Strategy', `Stress-watch: ${result.shock.driver} ${result.shock.direction} ${result.shock.magnitude}${result.shock.unit}`],
    ['Pair', pairs],
    ['Amount per Cycle', 'Paper only in AlphaArena — set your Playbook paper size'],
    ['Frequency', result.duration],
    ['Entry Condition', `Review when driver moves ${result.shock.magnitude}${result.shock.unit} ${result.shock.direction}; most sensitive: ${top[0]?.symbol || 'rNVDA'}`],
    ['Pause Condition', `Pause if realized stress exceeds worst estimated ${worst.toFixed(2)}% or invalidation triggers`],
    ['Market Regime', result.shock.category],
    ['Signal Confidence', `${result.model_source} | calibrated=${top.some((i) => i.calibrated)}`],
    ['Optimizer Reason', (result.explanation_view?.impact_summary || result.explanation || '').slice(0, 400)],
  ];
  return lines.map(([k, v]) => `${k}: ${v}`).join('\n');
}
